### ISSUE TITLE
[HTTP] http client is not properly URL encoding path

### ISSUE THREAD

**Issue by sanjiva** [ballerina-platform/ballerina-lang#19602]
It appears client->post("/a/b c/d", stuff) is not working correctly. Didn't test other actions but presumably its the same.

**Comment by chamil321** [comment]
@sanjiva, Actually the HTTP client does not perform encoding for a given path. In fact user has to provided the path after encoding it based on a chosen character set.
```ballerina
string encodedPath = check encoding:encodeUriComponent("/a/b c/d", "UTF-8");
var clientResponse = clientEP->post(encodedPath, req);
```
Should we improve the client to encode user given path before doing the outbound call? 

Required changes: Add charset as optional parameter
```ballerina
function post(@untainted string path, RequestMessage message, public string charset = "UTF-8") 
   returns Response|ClientError {
       
}
```

**Comment by sanjiva** [comment]
Yes IMO it needs to behave like the browser - the encoding always happens. I don't think there's a charset issue for URLs right??

Otherwise for computed URLs you always have to do this.

**Comment by chamil321** [comment]
When investigating possible scenarios, we understood that always encoding the path can lead to some misinterpretations in the URL. 

**Reserve characters might get encoded twice**

For example, if user needs to set a path with a reserved character, he should encode particular character during the URL formation to avoid the component conflicts according to the [spec](https://tools.ietf.org/html/rfc3986#section-2.2).

So user provided path as /date/s19%2F11%2F2019 (Here %2F stands for "/")

If the client encodes the path internally already encoded path, that violate the [spec](https://tools.ietf.org/html/rfc3986#section-2.4) as it clearly says that

> "Implementations must not
> percent-encode or decode the same string more than once, as decoding
> an already decoded string might lead to misinterpreting a percent
> data octet as the beginning of a percent-encoding, or vice versa in
> the case of percent-encoding an already percent-encoded string."

and also user cannot send the raw path as /date/s19/11/2019 as it is not equivalent path.

Hence, leaving the necessary encoding to perform by the URL creator, would keep the client behavior consistent IMO. 

However most of the browsers do the encoding for space(as mentioned in the issue), therefore we can special case the space and encode it always internally.
